python/tskit/combinatorics.py

```diff
@@ -0,0 +1,798 @@
+
+    @staticmethod
+    def from_range_rank(combination, n):
+        """
+        Find the combination of k integers from [0, n)
+        with the given rank in a lexicographic ordering.
+        """
+        k = len(combination)
+        if k == 0 or k == n:
+            return 0
+
+        j = combination[0]
+        combination = [x - 1 for x in combination]
+        if j == 0:
+            return Combination.from_range_rank(combination[1:], n - 1)
+
+        first_rank = Combination.comb(n - 1, k - 1)
+        rest_rank = Combination.from_range_rank(combination, n - 1)
+        return first_rank + rest_rank
+
+    @staticmethod
+    def unrank(rank, elements, k):
+        n = len(elements)
+        if k == 0:
+            return []
+
+        n_rest_combs = Combination.comb(n - 1, k - 1)
+        if rank < n_rest_combs:
+            return elements[:1] + Combination.unrank(rank, elements[1:], k - 1)
+
+        return Combination.unrank(rank - n_rest_combs, elements[1:], k)
+
+    @staticmethod
+    def with_replacement_rank(combination, n):
+        """
+        Find the rank of ``combination`` in the lexicographic ordering of
+        combinations with replacement of integers from [0, n).
+        """
+        k = len(combination)
+        if k == 0:
+            return 0
+        j = combination[0]
+        if k == 1:
+            return j
+
+        if j == 0:
+            return Combination.with_replacement_rank(combination[1:], n)
+
+        rest = [x - j for x in combination[1:]]
+        preceding = 0
+        for i in range(j):
+            preceding += Combination.comb_with_replacement(n - i, k - 1)
+        return preceding + Combination.with_replacement_rank(rest, n - j)
+
+    @staticmethod
+    def with_replacement_unrank(rank, n, k):
+        """
+        Find the combination with replacement of k integers from [0, n)
+        with the given rank in a lexicographic ordering.
+        """
+        if k == 0:
+            return []
+
+        i = 0
+        preceding = Combination.comb_with_replacement(n, k - 1)
+        while rank >= preceding:
+            rank -= preceding
+            i += 1
+            preceding = Combination.comb_with_replacement(n - i, k - 1)
+
+        rest = Combination.with_replacement_unrank(rank, n - i, k - 1)
+        return [i] + [x + i for x in rest]
+
+
+########################################################################
+# Helpers
+########################################################################
+
+
+def set_minus(arr, subset):
+    return [x for x in arr if x not in set(subset)]
+
+
+# TODO I think we can use part-count form everywhere. Right now
+# there's a janky work-around of grouping the partition when
+# we needed in part-count form but it doesn't look like there's any
+# place that can't just accept it from the start.
+def partition(n):
+    """
+    Ascending integer partitions of n.
+    """
+    if n > 0:
+        yield from rule_asc(n)
+
+
+def rule_asc(n):
```

Comment:
Let's leave rule_asc as it is, and do the dropping in your partition function. Should give some indication where this came from also, it'll look mysterious to readers otherwise.